db/python/layers/billing.py

```diff
@@ -60,87 +98,143 @@ async def query(
         billing_db = BillingDb(self.connection)
         return await billing_db.query(_filter, limit)
 
+    async def get_total_cost(
+        self,
+        fields: str,
+        start_date: str,
+        end_date: str,
+        topic: str | None = None,
+        cost_category: str | None = None,
+        sku: str | None = None,
+        ar_guid: str | None = None,
+        dataset: str | None = None,
+        batch_id: str | None = None,
+        sequencing_type: str | None = None,
+        stage: str | None = None,
+        sequencing_group: str | None = None,
+        order_by: str | None = None,
+        limit: int | None = None,
+        offset: int | None = None,
+    ) -> list[BillingTotalCostRecord] | None:
+        """
+        Get Total cost of selected fields for requested time interval
+        """
+        billing_db = BillingDb(self.connection)
+        return await billing_db.get_total_cost(
+            fields,
+            start_date,
+            end_date,
+            topic,
+            cost_category,
+            sku,
+            ar_guid,
+            dataset,
+            batch_id,
+            sequencing_type,
+            stage,
+            sequencing_group,
+            order_by,
+            limit,
+            offset,
+        )
+
 
 class BillingDb(BqDbBase):
```

Comment:
Agreed, although I believe that BigQuery also has a caching feature in place for repeat queries within a specified time frame. Did we want to look into caching both, the API cache or BQ cache? I suspect both would be ideal.

Reply:
I had checked the async-lru module, but it does not seem to cache properly, getting misses.

I had tested fastapi-cache2 module and that supports cache-control out of the box for GETs on the clients side properly, so if in the same browser session it will reuse. I had added to all our billing GET api points. Will be committing this change soon.
And yes BQ is caching the same queries, esp when using materialised views as we do.
I am thinking to add logging for all the BQ queries and store them in some kind of table together with time taken, query, user and cost (in MB billed), then we can revisit our API and optimise those often used or expensive.